genesyscloud/knowledge_document_variation/resource_genesyscloud_knowledge_document_variation.go

```diff
@@ -0,0 +1,349 @@
+package knowledgedocumentvariation
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+	"terraform-provider-genesyscloud/genesyscloud/consistency_checker"
+	"terraform-provider-genesyscloud/genesyscloud/provider"
+	"terraform-provider-genesyscloud/genesyscloud/util"
+	"terraform-provider-genesyscloud/genesyscloud/util/constants"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/retry"
+
+	resourceExporter "terraform-provider-genesyscloud/genesyscloud/resource_exporter"
+
+	knowledgeDocument "terraform-provider-genesyscloud/genesyscloud/knowledge_document"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/mypurecloud/platform-client-sdk-go/v150/platformclientv2"
+)
+
+type resourceIDs struct {
+	variationID         string
+	knowledgeBaseID     string
+	documentID          string
+	knowledgeDocumentID string
+}
+
+func getAllKnowledgeDocumentVariations(ctx context.Context, clientConfig *platformclientv2.Configuration) (resourceExporter.ResourceIDMetaMap, diag.Diagnostics) {
+	resources := make(resourceExporter.ResourceIDMetaMap)
+	variationProxy := getVariationRequestProxy(clientConfig)
+	knowledgeDocumentProxy := knowledgeDocument.GetKnowledgeDocumentProxy(clientConfig)
+
+	// Get all the Knowledge Base Entities
+	knowledgeBaseList, err := getAllKnowledgeBases(ctx, variationProxy)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, knowledgeBase := range knowledgeBaseList {
+		variationEntities, response, err := knowledgeDocumentProxy.GetAllKnowledgeDocumentEntities(ctx, &knowledgeBase)
+		if err != nil {
+			return nil, util.BuildAPIDiagnosticError(ResourceType, fmt.Sprintf("%v", err), response)
+		}
+
+		// Retrieve the documents for each knowledge base
+		for _, knowledgeDoc := range *variationEntities {
+			// parse document state
+			var documentState string
+
+			if knowledgeDoc.State != nil && *knowledgeDoc.State != "" {
+				isValidState := strings.EqualFold(*knowledgeDoc.State, "Published") || strings.EqualFold(*knowledgeDoc.State, "Draft")
+				if isValidState {
+					documentState = *knowledgeDoc.State
+				}
+			}
+
+			// get the variations for each document
+			knowledgeDocumentVariations, resp, getErr := variationProxy.getAllVariations(ctx, *knowledgeBase.Id, *knowledgeDoc.Id, documentState, nil)
+			if getErr != nil {
+				return nil, util.BuildAPIDiagnosticError(ResourceType, fmt.Sprintf("Failed to get page of knowledge document variations error: %v", getErr), resp)
+			}
+
+			if knowledgeDocumentVariations == nil || len(*knowledgeDocumentVariations) == 0 {
+				break
+			}
+
+			for _, knowledgeDocumentVariation := range *knowledgeDocumentVariations {
+				id := fmt.Sprintf("%s %s %s", *knowledgeDocumentVariation.Id, *knowledgeDoc.KnowledgeBase.Id, *knowledgeDoc.Id)
```

Comment:
This has bugged the heck out of me whenever I've happened across this particular resource. The building of the ID is not consistent. It seems like it should be knowledge base ID + knowledge doc ID + knowledge variation ID to align with the hierarchy of the parent resources.